add cli cmd for externally-signed tx

paweljakubas opened the pull request:
Add `cmdTransactionSubmit` to CLI of jormungandr main.

# Issue Number
#628 

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added CLI support for external tx
- [x] I have added respective integration tests for Jormungandr
- [x] I have added CLISpec unit tests for cmd
- [x] CLI accepts hex-encoded bytes to comply with what jcli returns
- [x] I transited from `json` to `octet-stream` content type  


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
add cli cmd for externally-signed tx

add cmdExternalTransaction to addCli of jormungandr main

paweljakubas committed:
add jormungandr integration tests

paweljakubas committed:
accept hex-encoded binary blob and add CLISpec unit tests

paweljakubas committed:
use Johannes proposal and add more integration test to CLI

correct swagger

paweljakubas committed:
add octet-stream support and corresponding tests

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -706,6 +706,12 @@ instance LiftHandler ErrSubmitExternalTx where
             , errReasonPhrase = errReasonPhrase err404
             }
+        ErrSubmitExternalTxWrongPayloadEncoding ->
+            apiError err404 MalformedTxPayload $ mconcat
```
Not about this PR, but some times we use multiline strings (`" \\n\ "`) and sometimes `mconcat`. Why? Should we have a preference?

paweljakubas replied:
yeah, we need to decide and be consistent throughout all codebase

KtorZ replied:
@Anviking we usually use `mconcat` when there are variables to include in the text formatting. When it's only plain text, the multiline notation works well. Though, we could probably use `mconcat` everywhere, I'd actually expect GHC to be "smart enough" to transform that correctly into a literal at compile-time. If not, the overhead in this situation is really nothing.

Anviking commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -266,9 +267,14 @@ errMsg403TxTooBig n = "I had to select " ++ show n ++ " inputs to construct the\
 
+errMsg404WronglyEncodedTxPayload :: String
+errMsg404WronglyEncodedTxPayload =
+    "I couldn't proceed with the payload due to wrong bytes\
```
- Is it worth it to say `proceed with the payload`? (Honest question)
- I think `bytes encoding` should be `byte encoding`.

>  Make sure your payload

The CLI help text says `BINARY_BLOB              hex-encoded binary blob`. Are there reasons for the various terminology (I could imagine this being referred to as tx too, not sure if it is), or can we opt for fewer terms?

What about something shorter like: `I couldn't decode the payload. It doesn't appear to be hex-encoded.` or something in that direction?

jonathanknowles replied:
@Anviking wrote:
> What about something shorter like: "I couldn't decode the payload. It doesn't appear to be hex-encoded. or something in that direction?"

I also think this wording is preferable.

paweljakubas replied:
corrected

Anviking commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -266,9 +267,14 @@ errMsg403TxTooBig n = "I had to select " ++ show n ++ " inputs to construct the\
 errMsg404MalformedTxPayload = "I couldn't decode the payload that seems to be an\
     \ externally-signed transaction due to: wrongly constructed binary blob. \
-    \Make sure to send a base64-encoded binary blob, in the proper binary \
+    \Make sure to send a hex-encoded binary blob, in the proper binary \
```
After seeing `proceed with the payload` below, I thought of a different direction here:

*I couldn't verify that the payload has the correct binary format. I won't send it to the node. Make sure the blob has the correct binary format (and is encoded as hex)*

paweljakubas replied:
sounds good. corrected to that

Anviking reviewed (commented):
Some thoughts on some error-messages

jonathanknowles commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -594,6 +595,27 @@ cmdTransactionList = command "list" $ info (helper <*> cmd) $ mempty
             (ApiT <$> mOrder)
 
+
+-- | Arguments for 'transaction create' command
```
Comment should refer to the `transaction submit` command.

paweljakubas replied:
corrected

jonathanknowles commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -594,6 +595,27 @@ cmdTransactionList = command "list" $ info (helper <*> cmd) $ mempty
+        runClient wPort Aeson.encodePretty $
+            postExternalTransaction (walletClient @t) wPayload
+
+
```
Too many blank lines here.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -1218,6 +1219,18 @@ listTransactionsViaCLI ctx args = cardanoWalletCLI @t $ join
+    , args
+    ]
+
+
```
Too many blank lines here.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
corrected

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -602,6 +603,10 @@ instance FromText a => FromHttpApiData (ApiT a) where
+instance MimeUnrender OctetStream PostExternalTransactionData where
+    mimeUnrender _ = pure . PostExternalTransactionData . BL.toStrict
+
+
```
Too many blank lines here.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

KtorZ replied:
It doesn't seems right that the `OctetStream` is hex-encoded here. The binary data would be better off raw and untouched :+1:

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -748,7 +750,8 @@ instance Arbitrary PostExternalTransactionData where
         count <- choose (0, 32)
         bytes <- BS.pack <$> replicateM count arbitrary
-        return $ PostExternalTransactionData bytes
+        let hex = convertToBase Base16 bytes :: B8.ByteString
```
Should we remove this type annotation?

paweljakubas replied:
indeed! removed the annotation

rvl replied:
I think the `newtype PostExternalTransactionData` wrapper can be removed altogether from the API and just replaced with ByteString.

In the CLI, you don't need `FromText PostExternalTransactionData`, just use [`eitherReader`](http://hackage.haskell.org/package/optparse-applicative-0.15.0.0/docs/Options-Applicative.html#v:eitherReader):

```haskell
argument (eitherReader (bimap show S8.unpack . convertFromBase Base16 . S8.pack))
```

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -184,7 +196,45 @@ spec = do
+                encodeTx (tx, wits) MsgTypeTransaction
+        let headers = Headers [ ("Content-Type", "application/octet-stream") ]
+        r <-
+            request @ApiTxId ctx postExternalTxEp headers (NonJson encodedSignedTx)
```
The line above is **too long** according to our coding standards.

Recommend:
```hs
        r <- request
            @ApiTxId ctx postExternalTxEp headers (NonJson encodedSignedTx)
```

paweljakubas replied:
corrected

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -184,7 +196,45 @@ spec = do
+                BL.fromStrict $ T.encodeUtf8 $
+                encodeWronglyTx (tx, wits) MsgTypeTransaction
+        r1 <-
+            request @ApiTxId ctx postExternalTxEp headers (NonJson wronglyEncodedTx)
```
Again, too long. Recommend:
```hs
        r1 <- request
            @ApiTxId ctx postExternalTxEp headers (NonJson wronglyEncodedTx)
```

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -184,7 +196,45 @@ spec = do
+            , expectResponseCode HTTP.status404
+            ]
+
+
```
Too many blank lines here.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+                              any (\(AddressAmount _ amnt) ->
+                                       amnt == Quantity amt)
+                              $ NE.toList $ tx ^. #outputs
+                         ) txsSrc
```
The above section of code uses variable-length indentation, but this is disallowed by our coding standards:

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-variable-length-indentation

Recommend:
```hs
    let txSrc:_ = flip filter txsSrc $ \tx -> any
            (\(AddressAmount _ amnt) -> amnt == Quantity amt) 
            (NE.toList $ tx ^. #outputs)
```

paweljakubas replied:
corrected

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+    let (ix, out) = if amnt1 == Quantity amt then
+            (0::Word32, out1)
+            else
+            (1::Word32, out2)
```
This formatting is weird, because the `else` keyword is on its own line, and is further to the left than the `if`.

I would recommend:
```hs
    let (ix, out) =
        if amnt1 == Quantity amt
        then (0::Word32, out1)
        else (1::Word32, out2)
```

paweljakubas replied:
corrected

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+            else
+            (1::Word32, out2)
+    let outAddr = getApiT $ fst $ API.address out
+    let myRnps = [(TxIn txHash ix, TxOut outAddr (Coin 100))]
```
Where does this `100` come from? Is it supposed to be the same value as `amt` above?

paweljakubas replied:
indeed, it should be `amt` defined above. corrected

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+    let left = if amt >= (fromIntegral feeMin) + toSend then
+            fromIntegral $ amt - (fromIntegral feeMin) - toSend :: Word64
+               else
+            0 :: Word64
```
This formatting is also odd, because the `if` expression is further to the left than the `if` keyword, and the `else` expression is further to the left than the `else` keyword. 

Would recommend this instead:
```hs
    let left =
            if amt >= (fromIntegral feeMin) + toSend
            then fromIntegral $ amt - (fromIntegral feeMin) - toSend :: Word64
            else 0 :: Word64```
```

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+               else
+            0 :: Word64
+    let txOuts = [ TxOut addrD (Coin left)
+                 , TxOut destAddr (Coin (fromIntegral toSend))]
```
This is another case of variable-length indentation, which is disallowed by the coding standards:

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-variable-length-indentation

Recommend:
```hs
    let txOuts =
            [ TxOut addrD (Coin left)
            , TxOut destAddr (Coin (fromIntegral toSend)) ]
```

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+                 , TxOut destAddr (Coin (fromIntegral toSend))]
+    let inps = fmap (second coin) myRnps
+    let block0H = unsafeFromHex
+            "301b1c634aa7b586da7243dd66a61bde904bc1755e9a20a9b5b1b0064e70d904"
```
Instead of copying and pasting the value here, can we reuse the value defined in `lib/jormungandr/test/integration/Main.hs`?

Perhaps that value can be moved into a shared constants file, or something similar, to avoid duplication.

paweljakubas replied:
done, moved `block0H` to Faucet and reuse in `Main` and `TransactionsSpec`

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+            $ putSignedTx inps' outs' wits'
+        toHex = convertToBase Base16 :: ByteString -> B8.ByteString
+    in T.decodeUtf8 $ toHex $ BL.toStrict $
+       encode (tx, wits) msgType
```
The above code uses variable-length alignment, this time by 3 characters.

Recommend this:
```hs
encodeTx (tx, wits) msgType =
    T.decodeUtf8 $ toHex $ BL.toStrict $ encode (tx, wits) msgType
  where
    encode ((Tx _ inps' outs'), wits') msgType' = runPut
        $ withHeader msgType'
        $ putSignedTx inps' outs' wits'
    toHex :: ByteString -> B8.ByteString
    toHex = convertToBase Base16
```

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-variable-length-indentation

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +294,114 @@ spec = do
+            $ withHeader msgType'
+            $ putSignedTx inps' outs' wits'
+    in T.decodeUtf8 $ convertToBase Base64 $ BL.toStrict $
+       encode (tx, wits) msgType
```
Variable-length alignment. Recommend:

```hs
encodeWronglyTx :: (Tx, [TxWitness]) -> MessageType -> Text
encodeWronglyTx (tx, wits) msgType =
    T.decodeUtf8 $ convertToBase Base64 $ BL.toStrict $
        encode (tx, wits) msgType
  where 
    encode ((Tx _ inps' outs'), wits') msgType' =
        runPut
            $ withHeader msgType'
            $ putSignedTx inps' outs' wits'
```

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-variable-length-indentation

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:
```diff
@@ -0,0 +1,85 @@
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:
```diff
@@ -0,0 +1,85 @@
+
+
+spec :: forall t. (EncodeAddress t, DecodeAddress t, KnownCommand t)
+     => SpecWith (Context t)
```
Variable-length indentation. Recommend:

```hs
spec
    :: forall t. (EncodeAddress t, DecodeAddress t, KnownCommand t)
    => SpecWith (Context t)     
```

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-variable-length-indentation

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:
```diff
@@ -0,0 +1,85 @@
+        out1 `shouldBe` ""
+        code1 `shouldBe` ExitFailure 1
+
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles reviewed (changes requested):
Looks mostly good! I would like to propose a small number of points to address:

### Names

Given that we're using `transaction submit` for the CLI command, I think we should also follow the existing conventions within the CLI module, where functions and data types are named consistently with the command with which they're associated. So, perhaps we should make the following renamings:

* `cmdExternalTransactionSend` -> `cmdTransactionSubmit`
* `ExternalTransactionSendArgs` -> `TransactionSubmitArgs`
* `payloadArgument` -> `transactionSubmitPayloadArgument`

### Error messages

I would second @Anviking's suggestion to use something like:
> "I couldn't decode the payload. It doesn't appear to be hex-encoded."

or:
> "I couldn't decode the attached payload. Please make sure that it is hex-encoded."

### Shared Constants

There are some constants which could be given names, or otherwise shared. A good example is `block0h` which appears to be the same as defined in `lib/jormungandr/test/integration/Main.hs`.

Perhaps these common constants should be turned into top-level definitions and then imported where needed, rather than copying and pasting?

### Coding Standards

There are quite a few parts of this PR that don't comply with our coding standards. Mainly there are three types of issue:

* Lines that are too long.
  https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters
* Variable-length indentation.
  https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-variable-length-indentation
* Extraneous vertical space.
  https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

Sorry to ask (I know it's a pain!), but would it be possible to go through the PR and check that you're following the standards, wherever possible? I've added comments wherever I spotted code that doesn't meet the standards, along with suggestions.

Other than the above points, looks good!

Jonathan

piotr-iohk commented on lib/core/src/Cardano/Wallet/Api.hs:
```diff
@@ -151,7 +152,7 @@ type ListTransactions t = "wallets"
 -- | https://input-output-hk.github.io/cardano-wallet/api/#operation/postExternalTransaction
 type PostExternalTransaction = "external-transactions"
-    :> ReqBody '[JSON] PostExternalTransactionData
+    :> ReqBody '[JSON, OctetStream] PostExternalTransactionData
```
What is the rationale/benefit of having support for both types, couldn't we just have either `Json` or `octet-stream`? (just asking because it basically means more testing)

jonathanknowles replied:
> What is the rationale/benefit of having support for both types, couldn't we just have either `Json` or `octet-stream`? (just asking because it basically means more testing)

If there's no clear requirement to support both, then I would also be in favour of cutting this down to just one.

Perhaps limit this to whichever option is easiest for the client.

paweljakubas replied:
I was thinking about it and initially I was persuaded to choose one way. Now, I am inclined to support both. Json is standard in our api, all users (on all technical levels) know how to use it. But on the other hand, more technical users will probably favor `octet-stream`. Frankly speaking when testing, I started to choose this content type. So for the time being,  I propose to leave it as it is (both content types reuse >95% of code), and decide whether leave it or nuke one type after confirming it with product manager (my understanding was we want two types)

KtorZ replied:
I don't think it's wise to support both here. JSON is a bit awkward as we are really just sending plain bytes, so `OctetStream` is more suitable. If any of the two has to prevail, I'd strongly suggest to keep `OctetStream`.

paweljakubas committed:
comply with PR review

hlint

add proper signing integration test and cleaning the code

comply with Rodney's suggestions

Matthias review changes

final fixes

Update swagger doc

last touch

rvl reviewed (dismissed):
Supplying the tx as a hex string on the command line seems like a workable solution for users.

But the changes to the API seem off.
- The bytestring inside `PostExternalTransactionData` should be raw tx bytes, not hex.
- The point of the non-json endpoint is that the user does not need to encode their tx in any way. They could just do `curl -H "Content-Type: application/octet-stream" --data-binary @mytx.bin http://localhost:8090/v2/external-transactions`

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +252,153 @@ spec = do
+                fixtureExternalTx ctx toSend
+        let encodedSignedTx = encodeTx txWits MsgTypeTransaction
+        r <-
+            request @ApiTxId ctx postExternalTxEp headers (payload encodedSignedTx)
```
This line is still too long!

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

paweljakubas replied:
oh, thanks for that. I have just installed 80-character line to my emacs config and see this. Should not have problems like that in the future

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +252,153 @@ spec = do
-        let wits = [sign bs result]
-        return (Tx (fragmentId inps txOuts wits) inps txOuts, wits)
+        let wronglyEncodedTx = encodeWronglyTx txWits MsgTypeTransaction
+        r1 <- request @ApiTxId ctx postExternalTxEp headers (payload wronglyEncodedTx)
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

paweljakubas replied:
corrected

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -293,99 +252,34 @@ spec = do
-    let txOuts =
-            [ TxOut addrD (Coin left)
-            , TxOut destAddr (Coin (fromIntegral toSend))
+        let wronglyEncodedTx = encodeWronglyTx txWits MsgTypeTransaction
```
It seems that `postExternalTxTest` has two tests inside: positive and negative. It would be better to have this in two separate test cases, so it is clear what actually fails (in case it fails)

paweljakubas replied:
I encapsulated both tests under one hood only for the sake of time savings. We are creating two wallets, there are numbers of calls, and thought that we can maybe put them together to save precious time

paweljakubas replied:
saying differently I aimed to call`fixtureExternalTx` as little times as possible - it is quite time consuming

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -594,6 +595,25 @@ cmdTransactionList = command "list" $ info (helper <*> cmd) $ mempty
+-- | Arguments for 'transaction submit' command
+data TransactionSubmitArgs = TransactionSubmitArgs
+    { _port :: Port "Wallet"
+    , _payload :: PostExternalTransactionData
```
I suppose this should really be `ByteString` (as discussed on Slack)

paweljakubas replied:
I use : 
```
newtype PostExternalTransactionData = PostExternalTransactionData (payload :: ByteString}
```
in order to deliver a number of instances and not use bytestring instances

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -264,10 +265,16 @@ errMsg403TxTooBig n = "I had to select " ++ show n ++ " inputs to construct the\
+    \format (and is encoded as hex when sent by json or raw when sent by\
+    \ octet-stream)."
+
+errMsg404WronglyEncodedTxPayload :: String
```
Maybe it's just a naming issue here but why is the error `404` for what seems to be an invalid request? The semantic looks off.

paweljakubas replied:
will change to `400` (Bad Response)

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -897,11 +900,15 @@ newWalletLayer tracer bp db nw tl = do
-        withExceptT ErrSubmitExternalTxNetwork $ postTx nw txWithWit
-        return tx
+    _submitExternalTx payload =
+        case convertFromBase Base16 payload of
```
It seems wrong that this is happening _here_ (a.k.a in the wallet business logic). The fact that this binary data might be hex-encoded is only relevant at the edge of the system (a.k.a API). As for addresses, witnesses and other `ByteString` in the code, we always consider them "raw" with no particular encoding. 

If a conversion should happen, it should be at the API-level only, but here, that payload should already be what is then called `validPayload` (which could be invalid though if it doesn't correspond to a signed transaction).

paweljakubas replied:
corrected

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -593,6 +593,9 @@ instance EncodeAddress t => ToText (AddressAmount t) where
         toText coins <> "@" <> encodeAddress proxy addr
 
+instance FromText PostExternalTransactionData where
+    fromText = pure . PostExternalTransactionData . T.encodeUtf8
```
This is where a conversion from hex should happen, if anywhere.

paweljakubas replied:
done

rvl reviewed (approved):
LGTM. I tweaked the wording in the swagger spec. You might want to squash the git history when rebasing.

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +238,152 @@ spec = do
-                }
-        r1 <- request @(ApiTransaction t) ctx (postTxEp wSrc) Default payload
-        verify r1
+    postExternalTxTest ctx = do
```
It seems to be used in only one test case, perhaps there is no need to have it as separate method.

paweljakubas replied:
ok, you are right. corrected

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:
```diff
@@ -0,0 +1,117 @@
+import Test.Integration.Framework.TestData
+    ( errMsg400MalformedTxPayload, errMsg400WronglyEncodedTxPayload )
+import Test.Integration.Jormungandr.Scenario.API.Transactions
+    ( ExternalTxFixture (..), encodeTx, fixtureExternalTx )
```
I think all those should live `Test.Integration.Framework.DSL` rather than `Test.Integration.Jormungandr.Scenario.API.Transactions`

paweljakubas replied:
I will put `encodeTx` to Jormungandr binary, `fixtureExternalTx` and `ExternalTxFixture` to DSL

paweljakubas replied:
ok, I moved `encodeTx` but cannot do it with others as they rely on Jormungandr `Tx`

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:
```diff
@@ -0,0 +1,117 @@
+            postExternalTransactionViaCLI @t ctx [argWrong]
+        err1 `shouldContain` errMsg400WronglyEncodedTxPayload
+        out1 `shouldBe` ""
+        code1 `shouldBe` ExitFailure 1
```
I think this should be a separate test case. The test reads `TRANS_EXTERNAL_CREATE_01 - proper single output transaction and proper binary format` yet there is negative test inside with wrongly encoded payload.

paweljakubas replied:
fair argument. seprated

jonathanknowles commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -263,11 +264,16 @@ errMsg403TxTooBig n = "I had to select " ++ show n ++ " inputs to construct the\
+    "I couldn't verify that the payload has the correct binary \
+    \format due to: wrongly constructed binary blob. \
+    \I won't send it to the node. Make sure the blob has the correct binary \
+    \format."
```
This message seems to repeat itself a bit. How about:
```hs
errMsg400MalformedTxPayload =
    "I couldn't verify that the payload has the correct binary format. \
    \Therefore I couldn't send it to the node. Please check the format \
    \and try again."
```

paweljakubas replied:
done

Anviking replied:
More conciseness: 👍 

But:

> Therefore I **couldn't** send it to the node.

I'm not sure the wallet is _unable_ to send the payload to the node. It _could_ skip verifying the payload, and just forward it to the node, right?

There is a small but real possibility that the wallet's notion of a correct transaction gets out of sync with the node's.

With that in mind I think it would be more transparent to say that the wallet _chooses_ not to send the transaction to the node. `won't` is one way to achieve that.

Counterarguments: 
- Could argue that `couldn't` already contains that according-to-myself meaning
- Could argue that this distinction _should_ be hidden from users.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -668,11 +667,10 @@ instance LiftHandler ErrSignTx where
+            apiError err400 MalformedTxPayload $ mconcat
+                [ "I couldn't verify that the payload has the correct binary "
+                , "format due to: ", pretty err, ". I won't send it to the node"
+                , ". Make sure the blob has the correct binary format."
```
Recommend changing "blob" to "payload" for consistency with earlier usage.

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Cardano/Faucet.hs:
```diff
@@ -21,6 +24,12 @@ initFaucet :: IO Faucet
+    "301b1c634aa7b586da7243dd66a61bde904bc1755e9a20a9b5b1b0064e70d904"
+    -- ^ jcli genesis hash --input test/data/jormungandr/block0.bin
+
+
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Main.hs:
```diff
@@ -117,7 +118,9 @@ main = hspec $ do
-            (TransactionsJormungandr.spec @t)
+            (TransactionsAPIJormungandr.spec @t)
+        describe "Transactions CLI endpoint tests (Jormungandr specific)"
+            (TransactionsCLIJormungandr.spec @t)
```
Perhaps `TransactionsApiJormungandr` and `TransactionsCliJormungandr` are more readable than `TransactionsAPIJormungandr` and `TransactionsCLIJormungandr`.

(Abbreviation in the middle of a compound name.)

Admittedly, we don't have a standard for this.

paweljakubas replied:
changed to `TransactionsApiJormungandr` and `TransactionsCliJormungandr`

jonathanknowles commented on lib/jormungandr/test/integration/Main.hs:
```diff
@@ -195,23 +198,20 @@ cardanoWalletServer mlisten = do
     jormungandrUrl = BaseUrl Http "localhost" 8080 "/api"
-    block0H = Hash $ unsafeFromHex
-        "301b1c634aa7b586da7243dd66a61bde904bc1755e9a20a9b5b1b0064e70d904"
-        -- ^ jcli genesis hash --input test/data/jormungandr/block0.bin
```
:+1:

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -170,33 +173,24 @@ spec = do
-            }|]
-        r <- request @ApiTxId ctx postExternalTxEp Default payload
+        (ExternalTxFixture _ _ _ txWits) <-
+                fixtureExternalTx ctx toSend
```
Probably don't need to wrap here!

paweljakubas replied:
indeed!

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -748,7 +749,8 @@ instance Arbitrary PostExternalTransactionData where
         bytes <- BS.pack <$> replicateM count arbitrary
-        return $ PostExternalTransactionData bytes
+        let hex = convertToBase Base16 bytes
+        return $ PostExternalTransactionData hex
```
This seems incorrect The conversion from and to hexadecimal should only happen at the edge of the system, `PostExternalTransactionData` should only contain raw bytes.

Having said that... Is this arbitrary instance used anywhere? We are running JSON roundtrips here, but since the JSON instance has disappeared :thinking: ?

paweljakubas replied:
agree here, we do not need conversion to hex anymore as we dealt with this at the api level. removed it.

nevertheless, we need arbitrary here as it is used in test : 
```
 describe "pointless tests to trigger coverage for record accessors"
```

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -170,33 +173,24 @@ spec = do
+        let wronglyEncodedSignedTx = encodeTx txWits MsgTypeInitial baseOk
+        let payload =
+                NonJson $ BL.fromStrict $
+                (toRawBytes baseOk) wronglyEncodedSignedTx
```
For this one, perhaps it's possible to save a line?
```hs
        let payload = NonJson $ BL.fromStrict $
                (toRawBytes baseOk) wronglyEncodedSignedTx
```

paweljakubas replied:
done

jonathanknowles commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +238,152 @@ spec = do
+                fixtureExternalTx ctx toSend
+        let baseOk = Base64
+        let encodedSignedTx = encodeTx txWits MsgTypeTransaction baseOk
+        let payload = NonJson . BL.fromStrict . (toRawBytes baseOk)
```
Probably don't need braces here. Perhaps:

```hs
       let payload = NonJson . BL.fromStrict . toRawBytes baseOk
```

paweljakubas replied:
indeed!

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +238,152 @@ spec = do
+            ]
+
+        let baseWrong = Base16
+        let wronglyEncodedTx = encodeTx txWits MsgTypeTransaction baseWrong
```
I acknowledge the comment about the time savings (https://github.com/input-output-hk/cardano-wallet/pull/664#discussion_r319909430) it seems that the test takes only ~4s (on my box) so I think it would be worth splitting it in two...

paweljakubas replied:
split the test into two

jonathanknowles reviewed (approved):
Looks mostly good to me!

I've made one suggestion to change the text of an error (see comments).

There are some minor formatting suggestions, but nothing serious.

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs:
```diff
@@ -244,101 +238,152 @@ spec = do
+    let password1 = "Secure Passphrase" :: Text
+    let createWallet = Json [json| {
+            "name": "Destination Wallet",
+            "mnemonic_sentence": #{mnemonics15},
```
Is it mandatory to use explicit `mnemonic15` as a destination wallet and then get the destination address from it? I'm wondering if we cannot just use `emptyWallet` method? 

The reason I'm asking is because when constructing tx using `jcli` you only need to provide destination address to the transaction being constructed (no key related to destination address is needed), so it seems the mnemonics don't have to be explicit (?)

paweljakubas replied:
cannot do it. I use something like that 
```
let (rootXPrv1, pwd1, st1) = getSeqState mnemonics15 password1
let (Just keysAddrDest) = isOwned st1 (rootXPrv1, pwd1) addrDest'
```
in order to populate keystore. I need well defined (ie. have password and mnemonic) wallet in order to sign it . When playing manually you can not be so restrictive in your test

piotr-iohk reviewed (commented):
I have few remarks about tests, however most of them I can probably address when working on further tests.

paweljakubas committed:
comply with Piotr remarks

move encodeTx back to Test.Integration.Jormungandr.Scenario.API.Transactions

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -593,6 +584,13 @@ instance EncodeAddress t => ToText (AddressAmount t) where
+instance FromText PostExternalTransactionData where
+    fromText text = case convertFromBase Base16 (T.encodeUtf8 text) of
+        Left _ ->
+            fail "Parse error. Expecting hex-encoded format."
```
Ok, so the API now expects raw bytes, and the CLI expects hex.

paweljakubas replied:
yes

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -645,3 +650,11 @@ decodeLegacyAddress bytes =
         <* CBOR.decodeWord32 -- CRC
+
+encodeTx :: (Tx, [TxWitness]) -> MessageType -> Base -> Text
+encodeTx (tx, wits) msgType base =
```
I don't think this function should live here.
- `Base`/`Text` conversion doesn't seem relevant to the Binary module
- `MessageType` should _always_ be `MsgTypeTransaction`? Why allow the caller to make mistakes?

I see you're using both these features in the tests, but I think the function should live there then.

piotr-iohk replied:
>I don't think this function should live here.

I thought it would go there `/lib/core-integration/src/Test/Integration/Framework/DSL.hs` ?

paweljakubas replied:
@Anviking fair point. I will put it back to `Test.Integration.Jormungandr.Scenario.API.Transactions` @piotr-iohk `encodeTx` use  `MessageType` which is Jormungandr specific, so we cannot move it in `core-integration` which as a rule should depend on `core` and our utilities.

Anviking reviewed (changes requested):
Looks good to me. But, 
- I definitely think the `encodeTx` helper should be moved

I haven't reviewed the integration test setup closely.
